numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -253,6 +256,13 @@ genint_type_str(PyObject *self)
     return item_str;
 }
 
+static PyObject *
+genbool_type_str(PyObject *self)
+{
+    PyBoolScalarObject *s = (PyBoolScalarObject *)self;
+    return _PyBool_ReprValues[s->obval];
```

Comment:
This hands out references without incrementing the reference count. You need to call `Py_INCREF(...)` on the value before returning it.

Note that this would be much easier to implement as:
```suggestion
    return PyUnicode_FromString(s->obval ? "numpy.True_", "numpy.False_");
```
although you'd lose the marginal benefits I mention in https://github.com/numpy/numpy/pull/17592/files#r508411504.